dash.el

```diff
@@ -2166,7 +2166,41 @@ See `-let' for the description of destructuring mechanism."
       ;; TODO: because inputs to the lambda are evaluated only once,
       ;; -let* need not to create the extra bindings to ensure that.
       ;; We should find a way to optimize that.  Not critical however.
-      `(lambda ,(--map (cadr it) inputs)
+      `(lambda ,(-map #'cadr inputs)
+         (-let* ,inputs ,@body))))))
+
+(defmacro -defun (name match-form &rest body)
+  "Return a `defun' NAME which destructures its input as MATCH-FORM and executes BODY.
+
+Note that you have to enclose the MATCH-FORM in a pair of parens,
+such that:
+
+  (-defun (x) body)
+  (-defun (x y ...) body)
+
+has the usual semantics of `defun'.  Furthermore, these get
+translated a into normal `defun', so there is no performance
+penalty.
+
+See `-let' for a description of destructuring mechanism."
+  (declare (doc-string 3) (indent defun)
+           (debug (&define name sexp
+                           [&optional stringp]
+                           [&optional ("declare" &rest sexp)]
+                           [&optional ("interactive" interactive)]
+                           def-body)))
+  (cond
+   ((nlistp match-form)
+    (signal 'wrong-type-argument (list #'listp match-form)))
+   ;; no destructuring, so just return regular defun to make things faster
+   ((-all? 'symbolp match-form)
+    `(defun ,name ,match-form ,@body))
+   (t
+    (let* ((inputs (--map-indexed (list it (make-symbol (format "input%d" it-index))) match-form)))
+      ;; TODO: because inputs to the defun are evaluated only once,
+      ;; -let* need not to create the extra bindings to ensure that.
+      ;; We should find a way to optimize that.  Not critical however.
```

Comment:
If this comment is not relevant to `-defun` then there is no point in cargo-culting it from `-lambda`.

Reply:
I can't tell if it is relevant or not, but if you or the maintainers want I could delete it.

Reply:
Hopefully @Fuco1 will enlighten us.

Reply:
Hopefully @Fuco1 will enlighten us.